### PR TITLE
Add option to restrict Google logins to custom domains

### PR DESCRIPTION
**Pull request by liggitt**

### CODE REVIEW

**Comment by liggitt** [comment]
@deads2k review

**Comment by liggitt** [comment]
@wshearn fyi

**Comment by wshearn** [comment]
Awesome work, thank you!

File: pkg/cmd/server/api/types.go [hunk 1]
```diff
@@ -284,6 +284,9 @@ type GoogleIdentityProvider struct {
 	ClientID string
 	// ClientSecret is the oauth client secret
 	ClientSecret string
+
+	// HostedDomain is the optional Google App domain to restrict logins to
```
Comment (deads2k) [inline pkg/cmd/server/api/types.go]: Add an example format in the godoc.

File: pkg/cmd/server/api/types.go [hunk 2]
```diff
@@ -301,6 +304,9 @@ type OpenIDIdentityProvider struct {
 	// ExtraScopes are any scopes to request in addition to the standard "openid" scope.
 	ExtraScopes []string
 
+	// ExtraAuthorizeParameters are any custom parameters to add to the authorize request.
+	ExtraAuthorizeParameters map[string]string
```
Comment (deads2k) [inline pkg/cmd/server/api/types.go]: What utility does this have for someone writing config?  I don't remember us stuffing all the claims into `.Extra` or anything, though I'm not necessarily opposed.  I would worry about staleness though.

Comment (deads2k) [inline pkg/cmd/server/api/types.go]: nm, parameters, not claims.

File: pkg/cmd/server/api/types.go [hunk 3]
```diff
@@ -284,6 +284,9 @@ type GoogleIdentityProvider struct {
 	ClientID string
 	// ClientSecret is the oauth client secret
 	ClientSecret string
+
+	// HostedDomain is the optional Google App domain to restrict logins to
+	HostedDomain string
```
Comment (deads2k) [inline pkg/cmd/server/api/types.go]: Is the format well known enough to validate well or would we be risking cutting off valid cases we don't anticipate?

Comment (liggitt) [inline pkg/cmd/server/api/types.go]: I couldn't find a description of the format, so I didn't want to assume standard subdomain format

**Comment by liggitt** [comment]
comments addressed

**Comment by deads2k** [comment]
lgtm [merge]

**Comment by liggitt** [comment]
docs in https://github.com/openshift/openshift-docs/pull/306

**Pull request merged**